src/measure_theory/stieltjes_measure.lean

```diff
@@ -0,0 +1,352 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Yury Kudryashov, Sébastien Gouëzel
+-/
+import measure_theory.borel_space
+
+/-!
+# Stieltjes measures on the real line
+
+Consider a function `f : ℝ → ℝ` which is monotone and right-continuous. Then one can define a
+corrresponding measure, giving mass `f b - f a` to the interval `(a, b]`.
+
+## Main definitions
+
+* `stieltjes_function` is a structure containing a function from `ℝ → ℝ`, together with the
+assertions that it is monotone and right-continuous. To `f : stieltjes_function`, one associates
+a Borel measure `f.measure`.
+* `f.left_lim x` is the limit of `f` to the left of `x`.
+* `f.measure_Ioc` asserts that `f.measure (Ioc a b) = of_real (f b - f a)`
+* `f.measure_Ioo` asserts that `f.measure (Ioo a b) = of_real (f.left_lim b - f a)`.
+* `f.measure_Icc` and `f.measure_Ico` are analogous.
+-/
+
+noncomputable theory
+open classical set filter
+open ennreal (of_real)
+open_locale big_operators ennreal nnreal topological_space
+
+/-! ### Basic properties of Stieltjes functions -/
+
+/-- Bundled monotone right-continuous real functions, used to construct Stieltjes measures. -/
+structure stieltjes_function :=
+(to_fun : ℝ → ℝ)
+(mono' : monotone to_fun)
+(right_continuous' : ∀ x, continuous_within_at to_fun (Ici x) x)
+
+namespace stieltjes_function
+
+instance : has_coe_to_fun stieltjes_function := ⟨_, to_fun⟩
+
+variable (f : stieltjes_function)
+
+lemma mono : monotone f := f.mono'
+
+lemma right_continuous (x : ℝ) : continuous_within_at f (Ici x) x := f.right_continuous' x
+
+/-- The limit of a Stieltjes function to the left of `x` (it exists by monotonicity). The fact that
+it is indeed a left limit is asserted in `tendsto_left_lim` -/
+@[irreducible] def left_lim (x : ℝ) := Sup (f '' (Iio x))
+
+lemma tendsto_left_lim (x : ℝ) : tendsto f (𝓝[Iio x] x) (𝓝 (f.left_lim x)) :=
+by { rw left_lim, exact f.mono.tendsto_nhds_within_Iio x }
+
+lemma left_lim_le {x y : ℝ} (h : x ≤ y) : f.left_lim x ≤ f y :=
+begin
+  apply le_of_tendsto (f.tendsto_left_lim x),
+  filter_upwards [self_mem_nhds_within],
+  assume z hz,
+  exact (f.mono (le_of_lt hz)).trans (f.mono h)
+end
+
+lemma le_left_lim {x y : ℝ} (h : x < y) : f x ≤ f.left_lim y :=
+begin
+  apply ge_of_tendsto (f.tendsto_left_lim y),
+  apply mem_nhds_within_Iio_iff_exists_Ioo_subset.2 ⟨x, h, _⟩,
+  assume z hz,
+  exact f.mono hz.1.le,
+end
+
+lemma left_lim_le_left_lim {x y : ℝ} (h : x ≤ y) : f.left_lim x ≤ f.left_lim y :=
+begin
+  rcases eq_or_lt_of_le h with rfl|hxy,
+  { exact le_rfl },
+  { exact (f.left_lim_le le_rfl).trans (f.le_left_lim hxy) }
+end
+
+/-- The identity of `ℝ` as a Stieltjes function, used to construct Lebesgue measure. -/
+protected def id : stieltjes_function :=
+{ to_fun := id,
+  mono' := λ x y, id,
+  right_continuous' := λ x, continuous_within_at_id }
+
+@[simp] lemma id_apply (x : ℝ) : stieltjes_function.id x = x := rfl
```

Comment:
```suggestion
@[simps] protected def id : stieltjes_function :=
{ to_fun := id,
  mono' := λ x y, id,
  right_continuous' := λ x, continuous_within_at_id }
```
You might also want an `initialize_simps_projections stieltjes_function (to_fun → apply)`.